runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,1963 @@
+			Some(s) => s,
+			None => return Err(Error::<T>::AncientDisputeStatement.into()),
+		};
+
+		let n_validators = session_info.validators.len();
+
+		// Check for ancient.
+		let (fresh, dispute_state) = {
+			if let Some(dispute_state) = <Disputes<T>>::get(&set.session, &set.candidate_hash) {
+				ensure!(
+					dispute_state.concluded_at.as_ref().map_or(true, |c| c >= &oldest_accepted),
+					Error::<T>::AncientDisputeStatement,
+				);
+
+				(false, dispute_state)
+			} else {
+				(
+					true,
+					DisputeState {
+						validators_for: bitvec![BitOrderLsb0, u8; 0; n_validators],
+						validators_against: bitvec![BitOrderLsb0, u8; 0; n_validators],
+						start: now,
+						concluded_at: None,
+					}
+				)
+			}
+		};
+
+		// Check and import all votes.
+		let summary = {
+			let mut importer = DisputeStateImporter::new(dispute_state, now);
+			for (statement, validator_index, signature) in &set.statements {
+				let validator_public = session_info.validators.get(validator_index.0 as usize)
+					.ok_or(Error::<T>::ValidatorIndexOutOfBounds)?;
+
+				// Check signature before importing.
+				check_signature(
+					&validator_public,
+					set.candidate_hash,
+					set.session,
+					statement,
+					signature,
+				).map_err(|()| Error::<T>::InvalidSignature)?;
+
+				let valid = match statement {
+					DisputeStatement::Valid(_) => true,
+					DisputeStatement::Invalid(_) => false,
+				};
+
+				importer.import(*validator_index, valid).map_err(Error::<T>::from)?;
+			}
+
+			importer.finish()
+		};
+
+		// Apply spam slot changes. Bail early if too many occupied.
+		let is_local = <Included<T>>::contains_key(&set.session, &set.candidate_hash);
+		if !is_local {
+			let mut spam_slots: Vec<u32> = SpamSlots::<T>::get(&set.session)
+				.unwrap_or_else(|| vec![0; n_validators]);
+
+			for (validator_index, spam_slot_change) in summary.spam_slot_changes {
+				let spam_slot = spam_slots.get_mut(validator_index.0 as usize)
+					.expect("index is in-bounds, as checked above; qed");
+
+				match spam_slot_change {
+					SpamSlotChange::Inc => {
+						ensure!(
+							*spam_slot < config.dispute_max_spam_slots,
+							Error::<T>::PotentialSpam,
+						);
+
+						*spam_slot += 1;
+					}
+					SpamSlotChange::Dec => {
+						*spam_slot = spam_slot.saturating_sub(1);
+					}
+				}
+			}
+
+			SpamSlots::<T>::insert(&set.session, spam_slots);
+		}
+
+		if fresh {
+			Self::deposit_event(Event::DisputeInitiated(
+				set.candidate_hash,
+				if is_local { DisputeLocation::Local } else { DisputeLocation::Remote },
+			));
+		}
+
+		{
+			if summary.new_flags.contains(DisputeStateFlags::FOR_SUPERMAJORITY) {
+				Self::deposit_event(Event::DisputeConcluded(
+					set.candidate_hash,
+					DisputeResult::Valid,
+				));
+			}
+
+			// It is possible, although unexpected, for a dispute to conclude twice.
+			// This would require f+1 validators to vote in both directions.
+			// A dispute cannot conclude more than once in each direction.
+
```

Comment:
For this to happen we need to have had equivocations in the dispute resolution protocol as well. Wondering whether we should make this more explicit to the client.

Reply:
I don't think we need to, because it's -EV for validators to equivocate anyway.